### PR TITLE
PostList: initial approach

### LINKED ISSUE #20545: Create a Proof of Concept Package for Deployment to all environments
#### Is your feature request related to a problem? Please describe.

We currently have features on WordPress.com, that are not available on Jetpack/Atomic sites, and some that have copies of code in Jetpack or other plugins, which creates problems with maintenance and parity of experience.

#### Describe the solution you'd like

We would like to explore creating features as composer packages within the Jetpack repo, so that we can then require them as dependencies in different environments. These should ease deployment and ensure that we share as much code as possible.

#### Describe alternatives you've considered

We have discussed creating separate plugins, or building on top of other plugins that we use for delivering functionality to all site types, but given the tooling that's already available in the Jetpack repo, this seems like the most appropriate place for development.

#### Additional context
There is some initial work on this concept that has been done for Nav Unification in #19726

### PR DESCRIPTION
**Pull request by retrofox**
This PR creates a Proof of Concept as an approach suggested to enhance the classic view of the WordPress admin, AKA, wp-admin.

### Share via package

It creates and exposes the whole implementation via a `automattic/jetpack-post-list` composer package (not published, yet)

It allows adding the dependency easily to the project :

```bash
composer require automattic/jetpack-post-list
```

### Use modern technologies in the UI

It uses React, Redux, etc. to rendering the UX, making the focus of the usage on @wordpress packages. It will be useful to try to keep the design parity with the block editor.

### Progressive enhancement

Although the vision of this package's goals could be super ambitious, we'd like to tackle the development/enhancement workflow in a progressively way, meaning that we could improve small sections and iterate later, again and again, reducing the risk of the introduction of errors, in favor of making the whole flow smoothly.

## Example: Posts list enhancement

This PR enhances the posts list page, specifically on the post date section.

### The approach

The app replaces the post date for each pos, into the posts list, rendering a React component that allows editing post date straight from there.
It uses the same components used in the settings section of the block editor (sidebar).

Block Editor | WPAdmin Plus / posts list page
---------|-----------
image | image

## Video
https://user-images.githubusercontent.com/77539/128521293-68ff151f-c694-4938-bf27-729018ed8e3e.mov

#### Changes proposed in this Pull Request:

*

### Testing instructions:

* checkout this branch

```bash
git checkout update/add-wp-admin-plus-package
```

* Jetpack dev-env . Add this dependency to Jetpack plugin

```bash
cd projects/plugins/jetpack && composer require automattic/jetpack-post-list
```

* wait for the dependencies update process. You should see something like the following

* Update pnpm just in case 🤞 

```bash
pnpm install -f
```

* Update Jetpack dependencies

```bash
jetpack install -r && jetpack install --all
```

* Run posts-list package compiling process

```bash
jetpack build packages/post-list
```

Use `watch` in case you'd like to dev editing the code in the client

```bash
jetpack watch packages/post-list
```

* Run web server

```
jetpack docker up
```

* Testing in a client / browser
* 
Load the posts admin page in the wp-admin, defining the `post-list=true` query variable:

`http://localhost/wp-admin/edit.php?post-list=true`

Confirm you can edit the post date by clicking on the date button:

## Tasks

- [x] Rename package to post-list. [more](https://github.com/Automattic/jetpack/pull/20589#issuecomment-896766000)
- [ ] Improve updating post date value. Consider do it when picker loses focus, `Undo` option, etc. [more](https://github.com/Automattic/jetpack/pull/20589#issuecomment-896766000)
- [ ] Show update message: when the date, status, etc changes. [more](https://github.com/Automattic/jetpack/pull/20589#issuecomment-896766000)
- [ ] Consider we might want to have a publish/draft toggle. [more](https://github.com/Automattic/jetpack/pull/20589#issuecomment-896766000)
- [ ] Improve experience resetting the calendar date value. [more](https://github.com/Automattic/jetpack/pull/20589#issuecomment-896766000)

### CODE REVIEW

**Comment by DustyReagan** [comment]
I got errors running `jetpack build packages/wp-admin-plus`.

Adding `"@babel/runtime": "7.14.0",` to the top of devDependencies in packages.json fixed the issue for me.

**Comment by retrofox** [comment]
> I got errors running `jetpack build packages/wp-admin-plus`.
> 
> Adding `"@babel/runtime": "7.14.0",` to the top of devDependencies in packages.json fixed the issue for me.

yeah, I thought that package was going to be required by [@wordpress/scripts](https://github.com/WordPress/gutenberg/blob/f9f6cf06a5d17fe2b593150bf75b1f37d4cb6910/packages/scripts/package.json#L34) but not sure, now.

File: projects/packages/wp-admin-plus/package.json [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+		"@wordpress/components": "15.0.0",
+		"@wordpress/core-data": "4.0.0",
+		"@wordpress/data": "5.2.0",
+		"@wordpress/date": "4.2.1",
+		"@wordpress/dom-ready": "3.2.1",
+		"@wordpress/editor": "10.2.0",
+		"@wordpress/element": "4.0.0",
+		"@wordpress/i18n": "4.2.1"
+	},
+	"devDependencies": {
```
Comment (DustyReagan) [inline projects/packages/wp-admin-plus/package.json:29]: I had to add "@babel/runtime": "7.14.0", to get this package working on my dev environment. It'd be worth getting someone else to install this package to see if they have the same issue.

**Review by DustyReagan — COMMENTED** [review]
Looks great! 👍 That dependency issue I ran into could just be my dev environment. 😅 But I think it's definitely worth getting another opinion before shipping.

File: projects/packages/wp-admin-plus/package.json (hunk 1 above)
Comment (allilevine) [reply projects/packages/wp-admin-plus/package.json:29]: I had the same issue, and the same fix worked. 👍

Comment (retrofox) [reply projects/packages/wp-admin-plus/package.json:29]: mmm, have tried updating `pnpm`?

Comment (allilevine) [reply projects/packages/wp-admin-plus/package.json:29]: `pnpm install -f` also fails, with `gyp: No Xcode or CLT version detected!`. I could try reinstalling Xcode.

Comment (allilevine) [reply projects/packages/wp-admin-plus/package.json:29]: I reinstalled Xcode and ran `pnpm install -f` successfully, but I still get the `Module not found: Error: Can't resolve '@babel/runtime/helpers/extends'` error.

Comment (retrofox) [reply projects/packages/wp-admin-plus/package.json:29]: :-/ let me add the package into the dependencies

**Comment by kraftbj** [comment]
In terms of initialization, we may want to use the Config package to ensure things aren't double-hooked (see https://github.com/Automattic/jetpack/tree/master/projects/packages/config ). That way if multiple plugins are including this feature, we're ensuring it is only loaded once.

cc: @Automattic/jetpack-kernel for an early review on adding this as a package, pitfalls, etc

**Comment by retrofox** [comment]
> In terms of initialization, we may want to use the Config package to ensure things aren't double-hooked (see https://github.com/Automattic/jetpack/tree/master/projects/packages/config ). That way if multiple plugins are including this feature, we're ensuring it is only loaded once.

Is it needed for a package, too? I haven't seen other packages are using this `automattic/jetpack-config` package.

**Comment by pablinos** [comment]
This is definitely a nice improvement to make, and it's great that we can reuse the components from the editor. Some high level thoughts while testing it:

- This is a nitpick, and you're definitely not going to like me for suggesting it, but I wonder if we should simplify the name. Maybe we could just call it `post-list` as it will contain enhancements for that table wherever it's used. Also `jetpack-` gets prepended to the name, so `jetpack-post-list` sounds kind of descriptive.
- Should we wait to update the date value? Maybe only do it when the picker loses focus? Alternatively, we could show a warning when the post is going to change status. The problem is that if someone is clicking around they might un-publish a post. It's a bit of an edge case, but could be worth considering. 
- Related to that we might want to have a publish/draft toggle. That way we can schedule a draft post to publish from the posts list.
- I'm not sure what the best value should be, but at the moment the reset link in the picker resets to the current date/time. It could be more useful to have that reset to the value when the picker was opened, or the value when the page was loaded. Maybe we could even give all three options, but that could lead to a cluttered interface.

**Comment by retrofox** [comment]
> * This is a nitpick, and you're definitely not going to like me for suggesting it

never

> ... , but I wonder if we should simplify the name. Maybe we could just call it `post-list` as it will contain enhancements for that table wherever it's used. Also `jetpack-` gets prepended to the name, so `jetpack-post-list` sounds kind of descriptive.

👍 

> * Should we wait to update the date value? Maybe only do it when the picker loses focus? Alternatively, we could show a warning when the post is going to change status. The problem is that if someone is clicking around they might un-publish a post. It's a bit of an edge case, but could be worth considering.

Yeah, we need to inform to the user about this significant change. Maybe we could experiment with an `Undo` action, too.

> * Related to that we might want to have a publish/draft toggle. That way we can schedule a draft post to publish from the posts list.

:100:

> * I'm not sure what the best value should be, but at the moment the reset link in the picker resets to the current date/time. It could be more useful to have that reset to the value when the picker was opened, or the value when the page was loaded. Maybe we could even give all three options, but that could lead to a cluttered interface.

Let me dig on it.

**Comment by kraftbj** [comment]
> > In terms of initialization, we may want to use the Config package to ensure things aren't double-hooked (see https://github.com/Automattic/jetpack/tree/master/projects/packages/config ). That way if multiple plugins are including this feature, we're ensuring it is only loaded once.
> 
> Is it needed for a package, too? I haven't seen other packages are using this `automattic/jetpack-config` package.

Yeah, else you can run into issues with it being called multiple times. 

In the consuming project, you'll want to `ensure` like this https://github.com/Automattic/jetpack/blob/master/projects/plugins/jetpack/class.jetpack.php#L823

and within the Config package, indicate what should happen when it is loaded, see this example for JITM ( https://github.com/Automattic/jetpack/blob/master/projects/packages/config/src/class-config.php#L164 ).

That should call whatever loader your package has to execute. 

Another alternative is simply to allow the consuming plugin (wpcomsh, the standalone plugin for this, etc) call the initiation routine.

In either event, a package shouldn't auto-execute it's own code upon loading, which this is doing with the `add_action`.

**Comment by jeherve** [comment]
Closing this for now because of the lack of activity on this. We can always reopen in the future if needed.

**Pull request closed without merging**